cpp/src/arrow/compute/kernels/vector_swizzle_test.cc

```diff
@@ -0,0 +1,781 @@
+                       MakeArrayOfNull(utf8(), max_integer - 1));
+  auto expected_suffix_value = values;
+  ASSERT_OK_AND_ASSIGN(auto expected,
+                       Concatenate({expected_prefix_nulls, expected_suffix_value}));
+  DoTestScatterAAA(values, indices, /*max_index=*/max_integer - 1, expected);
+}
+
+TEST(Scatter, Boolean) {
+  {
+    ARROW_SCOPED_TRACE("Basic");
+    auto values = "[true, false, true, true, false, false, true, true, true, false]";
+    auto indices = "[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]";
+    int64_t max_index = 9;
+    auto expected = "[false, true, true, true, false, false, true, true, false, true]";
+    TestScatter(boolean(), values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Values with nulls");
+    auto values = "[true, false, null, true, true, false, false, null, null, true]";
+    auto indices = "[9, 8, 7, 6, 5, 4, 3, 2, 1, 0]";
+    int64_t max_index = 9;
+    auto expected = "[true, null, null, false, false, true, true, null, false, true]";
+    TestScatter(boolean(), values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Indices with nulls");
+    auto values = "[true, false, true, true, false, false, true, true, true, false]";
+    auto indices = "[9, null, 7, null, 5, null, 3, null, 1, null]";
+    int64_t max_index = 9;
+    auto expected = "[null, true, null, true, null, false, null, true, null, true]";
+    TestScatter(boolean(), values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Output greater than input");
+    auto values = "[true, true, true, false, false, false]";
+    auto indices = "[0, 3, 6, 1, 4, 7]";
+    int64_t max_index = 8;
+    auto expected = "[true, false, null, true, false, null, true, false, null]";
+    TestScatter(boolean(), values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Values all null");
+    auto values = "[null, null]";
+    auto indices = "[0, 1]";
+    int64_t max_index = 1;
+    auto expected = "[null, null]";
+    TestScatter(boolean(), values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Indices all null");
+    auto values = "[true, false]";
+    auto indices = "[null, null]";
+    int64_t max_index = 1;
+    auto expected = "[null, null]";
+    TestScatter(boolean(), values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Empty input output null");
+    auto values = "[]";
+    auto indices = "[]";
+    int64_t max_index = 1;
+    auto expected = "[null, null]";
+    TestScatter(boolean(), values, indices, max_index, expected);
+  }
+  {
+    ARROW_SCOPED_TRACE("Indices duplicated indices");
+    auto values = "[true, false, null, null]";
+    auto indices = "[0, 1, 0, 1]";
+    int64_t max_index = 3;
+    auto expected = "[null, null, null, null]";
+    TestScatter(boolean(), values, indices, max_index, expected);
+  }
+}
+
+TEST(Scatter, Numeric) {
+  for (const auto& type : kSignedIntegerTypes) {
```

Comment:
Nit: call this ` value_type`?

Reply:
(also, why not test all number types: integers + floats?)

Reply:
Both addressed.

I was worrying our `ArrayFromJSON` isn't that robust (to parse integer into float). But it actually does the trick!